File: permutation2/solve.cpp
#include <iostream> 
#include <vector>

using namespace std;

class Solution {
public:
    bool compare(vector<int>& a, vector<int>& b) {
        for (int i = 0; i < a.size(); i++) {
            if (a[i] != b[i]) {
                return false;
            }
        }
        return true;
    }

    void backtracking(vector<int> &nums, vector<vector<int>> &res, vector<int> &save, vector<bool> &visited) {
        if (save.size() == nums.size()) {
            if (res.size() == 0) res.push_back(save);
            else {
                for (int i = 0; i < res.size(); i++) {
                    if (compare(res[i], save)) {
                        res[i] = save;
                        return;
                    }
                }
                res.push_back(save);
            }
            return;
        }
        for (int j = 0; j < nums.size(); j++) {
            if (visited[j]) continue;
            if (!visited[j]) {
                visited[j] = true;
                save.push_back(nums[j]);
                backtracking(nums, res, save, visited);
                save.pop_back();
                visited[j] = false;
            } 
        }
    }

    vector<vector<int>> permuteUnique(vector<int>& nums) {
        vector<vector<int>> res;
        vector<int> save;
        vector<bool> visited(nums.size(), false);
        backtracking(nums, res, save, visited);
        return res;
    }
};

int main() {
    vector<int> nums = {1, 1, 2};
    Solution s;
    vector<vector<int>> res = s.permuteUnique(nums);

    for (const auto& perm : res) {
        for (int i : perm) {
            cout << i << " ";
        }
        cout << endl;
    }

    return 0;
}

File: permutation2/solve.py
class Solution(object):
    def compare(self, a, b): 
        for i in range(len(a)):
            if a[i] != b[i]:
                return False
        return True
    
    def backtracking(self, nums, res, save, visited):
        if len(save) == len(nums):
            if not any(self.compare(r, save) for r in res):
                res.append(save[:])
            return
        
        for i in range(len(nums)):
            if visited[i]: continue
            visited[i] = True
            save.append(nums[i])
            self.backtracking(nums, res, save, visited)
            visited[i] = False
            save.pop()

    def permuteUnique(self, nums):
        res = []
        save = []
        visited = [False] * len(nums)
        self.backtracking(nums, res, save, visited)
        return res
    
if __name__ == "__main__":
    nums = [2, 2, 3]
    s = Solution()
    res = s.permuteUnique(nums)
    print(res)
